### PR TITLE
adding messaging plugin container with the various messaging plugins

### PR DESCRIPTION
**Pull request by ndefokou**
@Hermann-Core @Christiantyemele @IngridPuppet please review this PR

### CODE REVIEW

File: crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/handler.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    pub fn get_route(&self, name: &str) -> Result<Router, PluginContainerError> {
+        if self.loaded {
+            self.collected_routes
+                .get(name)
+                .cloned()
+                .ok_or(PluginContainerError::Unloaded)
+        } else {
+            Err(PluginContainerError::Unloaded)
+        }
+    }
```
Comment (Christiantyemele) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/handler.rs:147]: please can you give more comments to this function i think i understand what you are trying to do but not very sure, and you have provided no test for it too

File: crates/web-plugins/didcomm-messaging/protocol-plugins/forward/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "forward"
```
Comment (Christiantyemele) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/forward/Cargo.toml:2]: you can better use the protocols name here, use ```routing```

File: crates/web-plugins/didcomm-messaging/protocol-plugins/forward/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn it_works() {
+        let result = add(2, 2);
+        assert_eq!(result, 4);
+    }
+}
```
Comment (Christiantyemele) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/forward/src/lib.rs:14]: you can remove this test, i do not think its Adept having it in main

File: crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/plugins.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+use plugin_api::Plugin;
+
+lazy_static! {
+    pub(crate) static ref PLUGINS: Vec<Arc<Mutex<dyn Plugin + 'static>>> = vec![
+        #[cfg(feature = "protocol-forward")]
+        // Arc::new(Mutex::new(index::IndexPlugin {})),
+        #[cfg(feature = "protocol-pickup")]
+        // Arc::new(Mutex::new(did_endpoint::plugin::DidEndpoint {})),
+        #[cfg(feature = "protocole-mediator-coordination")]
```
Comment (Christiantyemele) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/plugins.rs:17]: it's preferable using underscores here instead of hyphens

**Review by Christiantyemele — CHANGES_REQUESTED** [review]
Nice Work Overall

File: crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/plugins.rs (hunk 4 above)
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/plugins.rs:17]: The hyphens are acceptable; we can proceed with as it is

File: crates/web-plugins/didcomm-messaging/protocol-plugins/routing/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "forward"
```
Comment (IngridPuppet) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/routing/Cargo.toml:2]: Shouldn't this name match the folder's name?

Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/routing/Cargo.toml:2]: It was an error, I renamed the folder but forgot to change the package name as well.

File: Cargo.toml [hunk 6]
```diff
@@ -15,6 +15,11 @@ members = [
     "crates/web-plugins/didcomm-messaging",
     "crates/web-plugins/oob-messages",
     "crates/web-plugins/didcomm-messaging/database",
+    "crates/web-plugins/didcomm-messaging/protocol-plugins/routing",
+    "crates/web-plugins/didcomm-messaging/protocol-plugins/handler",
+    "crates/web-plugins/didcomm-messaging/protocol-plugins/mediator-coordination",
+    "crates/web-plugins/didcomm-messaging/protocol-plugins/pickup",
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply Cargo.toml:21]: No, it can't be a cargo workspace itself; I tried that before, and it resulted in an error.

File: crates/web-plugins/didcomm-messaging/protocol-plugins/handler/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "handler"
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/handler/Cargo.toml:2]: like protocol-handler?

File: crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/handler.rs
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/handler.rs:1]: Because it is a component responsible for processing  messages

File: crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+pub enum PluginContainerError {
+    DuplicateEntry,
+    Unloaded,
+    PluginErrorMap(HashMap<String, PluginError>),
+}
+
+pub struct PluginContainer<'a> {
+    loaded: bool,
+    message_type_handlers: HashMap<String, Box<dyn Fn(&str) -> String + Send + Sync>>,
+    plugins: &'a Vec<Arc<Mutex<dyn Plugin>>>,
```
Comment (IngridPuppet) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs:18]: You consider message type handlers to have this signature:

```rust 
Fn(&str) -> String + Send + Sync
```

- How sure are you of this? Did you check the signature of current handlers?
- Also note that it is better to organize this concept around a plugin trait.
- How is this plugins property meaningful to you as is? `plugins: &'a Vec<Arc<Mutex<dyn Plugin>>>,`

File: crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+                tracing::error!("error mounting plugin {}", name);
+                errors.insert(name.clone(), err);
+            } else {
+                seen_names.insert(name.clone()); 
+    
+                tracing::info!("mounted plugin {}", name);
+                self.message_type_handlers.insert(
+                    format!("message.{}", name.clone()), 
+                    Box::new(move |msg: &str| format!("Handled by {}: {}", name, msg)) as Box<dyn Fn(&str) -> String + Send + Sync>
+                );
```
Comment (IngridPuppet) [inline crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs:76]: Can you provide an explanation of what is being done here?

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: crates/web-plugins/src/lib.rs
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/src/lib.rs:1]: No, that’s an error. I forgot to delete it. It was because I wanted to use * for importing

File: crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "handler"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+axum.workspace = true
+
+plugin-api.workspace = true
+dotenv-flow.workspace = true
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/Cargo.toml:10]: Yes axum and plugin-api but dotenv no

File: crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs [hunk 11]
```diff
@@ -0,0 +1,289 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, Mutex};
+
+use plugin_api::{Plugin, PluginError};
+
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs:5]: because it defines the contract that all plugins must follow i.e mounted, and unmounted.

File: Cargo.toml [hunk 12]
```diff
@@ -8,24 +8,25 @@ authors = ["adorsys GmbH Co. KG"]
...
-]
+    "crates/database",
+    "crates/web-plugins/didcomm-messaging/protocol-plugins/*",
 
+]
 
 [workspace.dependencies]
 did-utils = { path = "./crates/web-plugins/didcomm-messaging/did-utils", version = "0.1.0" }
-keystore = { path = "./crates/web-plugins/didcomm-messaging/keystore", version = "0.1.0" }
+keystore = { path = "./crates/keystore", version = "0.1.0" }
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply Cargo.toml:24]: No, it’s going to stay there because it is related to didcomm-messaging.

File: crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs (hunk 8 above)
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/protocol-plugins/protocole-handler/src/handler.rs:18]: - There is already a plugin trait, and when i am trying to add it is causing some other errors
-  The property plugins: &'a Vec<Arc<Mutex<dyn Plugin>>> indicates a reference to a vector of plugin instances wrapped in an Arc<Mutex<...>, allowing for shared ownership and thread-safe access. it  makes it possible to maintain and interact with multiple plugins concurrently.

**Comment by ndefokou** [comment]
@Hermann-Core, please review this pull request

File: crates/web-plugins/didcomm-messaging/message-api/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
+use std::{collections::HashMap, fmt::Debug, marker::PhantomData, sync::Arc};
+
+type MessageHandler<S, M, R> = fn(Arc<S>, M) -> Result<R, PluginError>;
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/message-api/src/lib.rs:3]: The handlers are asynchronous functions, thus this function pointer will not function correctly. The result in question is the error result from the handler functions and is not associated with the Plugin interface.
Regarding the `Ok` result, it must match the type of the function's second argument.

Concerning the arguments, it is unnecessary to add an `Arc` to the first argument because an `Arc` necessitates the `Clone`, `Sync`, and `Send` bounds, allowing for the safe removal of the `Arc`.

File: crates/web-plugins/didcomm-messaging/message-api/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn route(mut self, msg_type: &str, f: MessageHandler<S, M, R>) -> Self {
+        self.routes.insert(msg_type.to_string(), f);
+        self
+    }
+
+    pub fn merge(&mut self, other: &Self) {
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/message-api/src/lib.rs:39]: Ideally, this function should consume `self` and incorporate the other message router, allowing it to be chained and collect the other router within the same instance instead of invoking it multiple times.

File: crates/web-plugins/didcomm-messaging/protocols/forward/src/plugin.rs [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    fn mount(&mut self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn unmount(&self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn routes(&self) -> MessageRouter<S, M, R> {
+        MessageRouter::new()
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/protocols/forward/src/plugin.rs:24]: Where is/are the registered route/s for the forward protocol?

File: crates/web-plugins/didcomm-messaging/protocols/mediator-coordination/src/lib.rs [hunk 16]
```diff
@@ -3,3 +3,4 @@ pub mod web;
 
 mod jose;
 mod model;
+mod plugin;
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/protocols/mediator-coordination/src/lib.rs:6]: The plugin needs to be public to be accessible from the manager.

File: crates/web-plugins/didcomm-messaging/protocols/mediator-coordination/src/plugin.rs [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    fn mount(&mut self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn unmount(&self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn routes(&self) -> MessageRouter<S, M, R> {
+        MessageRouter::new()
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/protocols/mediator-coordination/src/plugin.rs:25]: Indeed, the `routes` function registers all the routes used by a **didcomm-messaging** protocol. You are providing an empty router here.

File: crates/web-plugins/didcomm-messaging/protocols/pickup/src/lib.rs [hunk 18]
```diff
@@ -2,3 +2,4 @@ mod error;
 mod model;
 
 pub mod handler;
+mod plugin;
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/protocols/pickup/src/lib.rs:5]: `plugin` need to be public.

File: crates/web-plugins/didcomm-messaging/protocols/pickup/src/plugin.rs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    fn mount(&mut self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn unmount(&self) -> Result<(), PluginError> {
+        Ok(())
+    }
+
+    fn routes(&self) -> MessageRouter<S, M, R> {
+        MessageRouter::new()
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/protocols/pickup/src/plugin.rs:25]: See the previous comments.

File: crates/web-plugins/didcomm-messaging/src/lib.rs [hunk 20]
```diff
@@ -1,6 +1,6 @@
 mod manager;
 mod midlw;
-mod protocols;
+mod protocol;
```
Comment (Hermann-Core) [inline crates/web-plugins/didcomm-messaging/src/lib.rs:3]: why did you rename from `protocols` to `protocol`?

**Review by Hermann-Core — CHANGES_REQUESTED** [review]

File: crates/web-plugins/didcomm-messaging/src/lib.rs (hunk 20 above)
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/src/lib.rs:3]: There were two folders of the same name.

Comment (Hermann-Core) [reply crates/web-plugins/didcomm-messaging/src/lib.rs:3]: > There were 1 folders and a file of the same name.

There should be only one module with that name. Where does the other one come from?

File: crates/web-plugins/didcomm-messaging/Cargo.toml [hunk 21]
```diff
@@ -15,15 +15,27 @@ forward.workspace = true
 pickup.workspace = true
 trust-ping.workspace = true
 mediator-coordination.workspace = true
```
Comment (IngridPuppet) [inline crates/web-plugins/didcomm-messaging/Cargo.toml:17]: Make these dependencies optional:

```toml
forward = { workspace = true, optional = true }
pickup = { workspace = true, optional = true }
trust-ping = { workspace = true, optional = true }
mediator-coordination = { workspace = true, optional = true }
```

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: src/plugins.rs [hunk 22]
```diff
@@ -18,6 +18,6 @@ lazy_static! {
         #[cfg(feature = "plugin-oob_messages")]
         Arc::new(Mutex::new(oob_messages::plugin::OOBMessages {})),
         #[cfg(feature = "plugin-didcomm_messaging")]
-        Arc::new(Mutex::new(didcomm_messaging::plugin::MediatorCoordination::default())),
+        Arc::new(Mutex::new(didcomm_messaging::plugin::DidcommMessaging::default())),
     ];
 }
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply src/plugins.rs:23]: Yes, it is. Yes, but it was not handled in this file. I saw it while working on this ticket and decided to modify it since I was the one who worked on that ticket.

File: crates/web-plugins/didcomm-messaging/Cargo.toml (hunk 21 above)
Comment (Hermann-Core) [reply crates/web-plugins/didcomm-messaging/Cargo.toml:17]: Yeah, it makes sense.

File: crates/web-plugins/didcomm-messaging/src/manager.rs [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+    loaded: bool,
+    collected_routes: Vec<MessageRouter>,
+    message_plugins: &'a Vec<Arc<dyn MessagePlugin>>,
+}
+
+impl<'a> MessagePluginContainer<'a> {
+    pub(crate) fn new() -> Self {
+        Self {
+            loaded: false,
+            collected_routes: vec![],
```
[the remark this answers is not in the document]
Comment (Hermann-Core) [reply crates/web-plugins/didcomm-messaging/src/manager.rs:23]: I forgot to add test here. @ndefokou can you provide tests here please?

File: crates/web-plugins/didcomm-messaging/src/web/dispatcher.rs [hunk 24]
```diff
@@ -1,80 +1,35 @@
...
-            .await
-            .map_err(|e| e.into_response()),
-
-        _ => return (StatusCode::BAD_REQUEST, "Unsupported operation".to_string()).into_response(),
-    };
+    let mut container = MessagePluginContainer::new();
+    if container.load().is_err() {
+        tracing::error!("failed to load DIDComm protocols container");
+        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
+    }
```
[the remark this answers is not in the document]
Comment (Hermann-Core) [reply crates/web-plugins/didcomm-messaging/src/web/dispatcher.rs:21]: My bad. let fix that.

**Review by IngridPuppet — APPROVED** [review]

File: crates/web-plugins/didcomm-messaging/src/manager.rs [hunk 25]
```diff
@@ -81,3 +65,123 @@ impl<'a> MessagePluginContainer<'a> {
...
+
+        let mut container = MessagePluginContainer {
+            loaded: false,
+            collected_routes: vec![],
+            message_plugins: &plugins,
+        };
+
+        assert!(container.load().is_ok());
+        assert_eq!(container.collected_routes.len(), 2);
+    }
```
[the remark this answers is not in the document]
Comment (ndefokou) [reply crates/web-plugins/didcomm-messaging/src/manager.rs:164]: Okay

**Review by IngridPuppet — APPROVED** [review]

File: crates/web-plugins/didcomm-messaging/src/plugin.rs [hunk 26]
```diff
@@ -85,26 +116,34 @@ impl Plugin for MediatorCoordination {
...
-                tracing::error!("Failed to load app state: {:?}", err);
-                PluginError::Other("Failed to load app state".to_owned())
-            })?;
+        let state = AppState::from(
+            env.public_domain.clone(),
+            diddoc,
+            Some(msg_types.clone()),
+            Some(repository),
+        )
+        .map_err(|err| PluginError::Other(format!("Failed to load app state: {:?}", err)))?;
```
Comment (Blindspot22) [inline crates/web-plugins/didcomm-messaging/src/plugin.rs:146]:
```suggestion
.map_err(|err| {
            error!("Failed to load app state: {:?}", err);
            PluginError::Other(format!("Failed to load app state: {:?}", err))
        })?;
```

File: crates/web-plugins/didcomm-messaging/src/plugin.rs [hunk 27]
```diff
@@ -85,26 +116,34 @@ impl Plugin for MediatorCoordination {
...
 
         // Build router
```
Comment (Blindspot22) [inline crates/web-plugins/didcomm-messaging/src/plugin.rs:148]:
```suggestion
info!("Routes built successfully.");
```

**Review by Hermann-Core — APPROVED** [review]

**Review by Christiantyemele — APPROVED** [review]

**Review by Blindspot22 — CHANGES_REQUESTED** [review]
*Suggestion*
Enhance the plugin with logging to improve observability during initialization and operation. You can use the `tracing crate` for structured logging as indicated in the review.

**Comment by Hermann-Core** [comment]
> _Suggestion_ Enhance the plugin with logging to improve observability during initialization and operation. You can use the `tracing crate` for structured logging as indicated in the review.

These are unnecessary and redundant because all the errors and required informations will already be traced. Take a look at the plugin manager implementation.
